examples/model-viewer/main.cpp

```diff
@@ -1018,13 +1018,13 @@ struct ShaderCache : RefObject
             return true;
         }
 
-        UInt GetHashCode() const
+        Slang::HashCode getHashCode() const
```

Comment:
It is probably overkill to try to add to this PR, but I'd like to see us eventually move to a model where hashable types implement something like:

```c++
void hashInto(Hasher& hasher);
```

And then there can be a single generic `getHashCode()` operation that is:

```c++
template<typename T>
HashCode getHashCode(T const& value)
{
    Hasher hasher;
    value.hashInto(hasher);
    return hasher.getResult();
}
```

The reason for the alternative design is that we currently have a lot of cases where an aggregate type computes its hash by calling `GetHashCode` recursively on its fields/contents and then using `combineHash` to reduce things down. This approach has two big drawbacks:

1. There's an annoying amount of boilerplate code because you need both the `GetHashCode()` and `combineHash()` calls for each field, and you often need to thread through a local variable like `hash` below, which gets uttered twice for each field.

2.  The quality of the hash is compromised in that when computing the hash of a type with many bits of state, you reduce each of its fields down to the number of bits in `HashCode` and *then* combine then, instead of combining them into an intermediate hash with a larger number of bits, and then reducing it down at the end.

Having a distinct `Hasher` type solves both of those problems:

1. Calling `hasher.hash(something)` can both handle the recursive hashing of `something` into the hasher, and deal with combining it with the hash state that was already in place.

2. Because the `Hasher` type is opaque and only used during the computation of hash codes, it can use more bits for its internal state than `HashCode` does.

I'm not arguing for this change to get added to this PR, but it might be worth considering as a follow-up if we are changing all these places in the code anyway.

Reply:
I mean following this logic, it would argue for having the virtual methods be not getHashCode but intoHasher also. Then the final, getting of the HashCode is only something that is done at the very end. 

A nice advantage of such a design is that it allows for more bits for the hash state, and that the final hash mixing can be a one off performed when the hash code is produced at the end.

I also agree the combineHash mechanism adds a lot of cruft.

You could also use operator<< overload on the hasher to allow doing 

```
void Thing::intoHasher(Hasher& hasher)
{
    hasher << m_fieldA << m_anotherField << m_andAnother;
}
```
```

Reply:
I think we are in violent agreement here.